### PR TITLE
Filters for Pair Indexers

### PR DESCRIPTION
**Pull request by sakibmatin**
PyTorch module to filter various pair_lists (`pair_dist`, `pair_first`, ...) by distance in `pair_first`. 
`hippynn/hippynn/layers/pairs/filters.py`

Graph Nodes `OpenFilter` and `PeriodicFilter`. Structure is similar to existing `OpenPairIndexer` and `PeriodicPairIndexer`
`hippynn/hippynn/graphs/nodes/pairs.py`

ASE graph construction automatically uses Filters whenever multiple PairIndexers are present in the model.  
`hippynn/hippynn/interfaces/ase_interface/calculator.py`  

**TODO**
- Doc Strings.
- Remove GraphViz code in `ase_interface/calculator.py`
- Unit Tests

### CODE REVIEW

File: hippynn/interfaces/ase_interface/calculator.py [hunk 1]
```diff
@@ -134,6 +153,20 @@ def setup_ASE_graph(energy, charges=None, extra_properties=None):
     check_link_consistency((*new_inputs, *implemented_nodes))
     mod = GraphModule(new_inputs, implemented_nodes)
     mod.eval()
+
+    ### Graph for Debugging (TODO Remove for final PR.)
```
Comment (lubbersnick) [inline hippynn/interfaces/ase_interface/calculator.py:157]: This is marked for removal, I guess it should be removed

File: hippynn/graphs/nodes/pairs.py [hunk 2]
```diff
@@ -426,51 +426,32 @@ def __init__(self, name, parents, module="auto", **kwargs):
...
-    def expand0(self, PI, purpose): 
-        return PI.pair_dist, PI.pair_first, PI.pair_second, PI.pair_coord
-
-    @_parent_expander.match(ExternalNeighborIndexer)
-    def expand0(self, PI, purpose):
-        return PI.pair_dist, PI.pair_first, PI.pair_second, PI.pair_coord
-
+    @_parent_expander.match(PairIndexer)
+    def expand0(self, pair_indexer, purpose):
+        parents = pair_indexer.children
```
[the remark this answers is not in the document]
Comment (lubbersnick) [reply hippynn/graphs/nodes/pairs.py:435]: Ok, this is an exciting development. Buried away is this code: https://github.com/lanl/hippynn/blob/main/hippynn/graphs/nodes/base/definition_helpers.py#L36

During initialization, every node is connected to its current set of parents. This enables the finding of related nodes by traversing the current temporary graph. (find_unique_relative and similar.) 

Because the node isn't finished building, it triggers this code. https://github.com/lanl/hippynn/blob/75d1e9ad66778cdb4336e462ba56cad7ee7e93eb/hippynn/graphs/nodes/base/node_functions.py#L136

What you supply will work currently, but I consider it a workaround because it depends on the implementation for temporary connections. We should try to find a more robust way to do this. A slightly more robust way would be to take only the children of the pair finder which are of type IndexNode. (Only thing that an IndexNode does is take out some piece of the result list, so it is almost universally associated with a MultiNode). There might also be some kind of fix to how the the temporary parents/children work or a tweak to a multinode.

[the remark this answers is not in the document]
Comment (lubbersnick) [reply hippynn/graphs/nodes/pairs.py:435]: Actually, rather than take only the IndexNodes from the children, maybe we should just take all children that are not "self"-that should do it I hope.

[the remark this answers is not in the document]
Comment (lubbersnick) [reply hippynn/graphs/nodes/pairs.py:435]: Another solution: Only take children which definitely have IdxType pair. Then the errors can't occur.
Not sure immediately which solution has the best consequences. One thing I am slightly unhappy with is that this node can only take PairFinder as its parents.
Yet another: the parents for this node could be explicitly the pairfinder children, so that we build it as ``PairFilter(name,pairfinder.children,radius)``.

[the remark this answers is not in the document]
Comment (sakibmatin) [reply hippynn/graphs/nodes/pairs.py]: I have put in a tentative fix that appears to be robust. 
By checking that all children in `pairfinder.children` have the attribute `_index_state`. This ensures that only properly initialized PairIndexers are extracted. This follows one of your suggestions in :

> Another solution: Only take children which definitely have IdxType pair. Then the errors can't occur. Not sure immediately which solution has the best consequence ....

This works when constructing the ASE Calculator for a Hipnn network with upto 3 PairIndexers. 

Not clear if there are any other edge cases missed by the check put in so far.

**Comment by sakibmatin** [comment]
Can be merged in. The Unit test for ASE graph construction may be overkill.
